auth/iam_ipa.go

```diff
@@ -0,0 +1,394 @@
+
+	ipa := ipaIAMService{
+		id:        0,
+		version:   IpaVersion,
+		host:      host,
+		vaultName: vaultName,
+		username:  username,
+		password:  password,
+		debug:     debug,
+		rootAcc:   rootAcc,
+	}
+	jar, err := cookiejar.New(nil)
+	if err != nil {
+		// this should never happen
+		panic(err)
+	}
+
+	mTLSConfig := &tls.Config{InsecureSkipVerify: isInsecure}
+	tr := &http.Transport{
+		TLSClientConfig: mTLSConfig,
+	}
+	ipa.client = http.Client{Jar: jar, Transport: tr}
+
+	err = ipa.login()
+	if err != nil {
+		return nil, err
+	}
+
+	req := ipa.newRequest("vaultconfig_show/1", []string{}, map[string]any{"all": true})
+	vaultConfig := struct {
+		Kra_Server_Server             []string
+		Transport_Cert                Base64EncodedWrapped
+		Wrapping_default_algorithm    string
+		Wrapping_supported_algorithms []string
+	}{}
+	_, err = ipa.rpc(req, &vaultConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	cert, err := x509.ParseCertificate(vaultConfig.Transport_Cert)
+	if err != nil {
+		return nil, err
+	}
+
+	ipa.kraTransportKey = cert.PublicKey.(*rsa.PublicKey)
+
+	isSupported := false
+	for _, algo := range vaultConfig.Wrapping_supported_algorithms {
+		if algo == "aes-128-cbc" {
+			isSupported = true
+			break
+		}
+	}
+
+	if !isSupported {
+		return nil, fmt.Errorf("IPA vault does not support aes-128-cbc. Only %v supported", vaultConfig.Wrapping_supported_algorithms)
+	}
+	return &ipa, nil
+}
+
+func (ipa *ipaIAMService) CreateAccount(account Account) error {
+	return fmt.Errorf("not implemented")
+}
+
+func (ipa *ipaIAMService) GetUserAccount(access string) (Account, error) {
+	if access == ipa.rootAcc.Access {
+		return ipa.rootAcc, nil
+	}
+
+	req := ipa.newRequest("user_show/1", []string{access}, map[string]any{})
+
+	userResult := struct {
+		Gidnumber []string
+		Uidnumber []string
+	}{}
+	_, err := ipa.rpc(req, &userResult)
+	if err != nil {
+		return Account{}, err
+	}
+
+	uid, _ := strconv.Atoi(userResult.Uidnumber[0])
+	gid, _ := strconv.Atoi(userResult.Gidnumber[0])
+	account := Account{
+		Access:  access,
+		Role:    RoleUser,
+		UserID:  uid,
+		GroupID: gid,
+	}
+
+	session_key := make([]byte, 16)
+	rand.Read(session_key)
+	encrypted_key, err := rsa.EncryptPKCS1v15(rand.Reader, ipa.kraTransportKey, session_key)
+	if err != nil {
+		return account, err
+	}
+	req = ipa.newRequest("vault_retrieve_internal/1", []string{ipa.vaultName},
+		map[string]any{"username": access,
+			"session_key":   Base64EncodedWrapped(encrypted_key),
+			"wrapping_algo": "aes-128-cbc"})
+	data := struct {
+		Vault_data Base64EncodedWrapped
+		Nonce      Base64EncodedWrapped
+	}{}
+	_, err = ipa.rpc(req, &data)
+	if err != nil {
+		return account, err
+	}
+
+	aes, _ := aes.NewCipher(session_key)
+	cbc := cipher.NewCBCDecrypter(aes, data.Nonce)
+	cbc.CryptBlocks(data.Vault_data, data.Vault_data)
+	secret_unpadded_json, _ := pkcs7Unpad(data.Vault_data, 16)
```

Comment:
is data ever populated with real values? it looks like rpc() doesnt actually modify data, and above is just creating these with 0 values?

Reply:
It does, it un-marshals value in rpc call. I updated that method signature to avoid confusion.